test_conformance/api/test_api_consistency.cpp

```diff
@@ -1151,3 +1151,148 @@ int test_consistency_3d_image_writes(cl_device_id deviceID, cl_context context,
+        test_error(error, "Unable to get device CL_DEVICE_DOUBLE_FP_CONFIG");
+        test_assert_error(value == 0,
+                          "CL_DEVICE_DOUBLE_FP_CONFIG must return 0");
+
+        error =
+            clGetDeviceInfo(deviceID, CL_DEVICE_PREFERRED_VECTOR_WIDTH_DOUBLE,
+                            sizeof(value), &value, nullptr);
+        test_error(
+            error,
+            "Unable to get device CL_DEVICE_PREFERRED_VECTOR_WIDTH_DOUBLE");
+        test_assert_error(
+            value == 0,
+            "CL_DEVICE_PREFERRED_VECTOR_WIDTH_DOUBLE must return 0");
+
+        error = clGetDeviceInfo(deviceID, CL_DEVICE_NATIVE_VECTOR_WIDTH_DOUBLE,
+                                sizeof(value), &value, nullptr);
+        test_error(error,
+                   "Unable to get device CL_DEVICE_NATIVE_VECTOR_WIDTH_DOUBLE");
+        test_assert_error(value == 0,
+                          "CL_DEVICE_NATIVE_VECTOR_WIDTH_DOUBLE must return 0");
+    }
+
+    return TEST_PASS;
+}
+
+int test_consistency_requirements_fp16(cl_device_id deviceID,
+                                       cl_context context,
+                                       cl_command_queue queue, int num_elements)
+{
+    cl_int error = CL_SUCCESS;
+    cl_device_fp_config value = 0;
+
+    if (is_extension_available(deviceID, "cl_khr_fp16"))
+    {
+        error = clGetDeviceInfo(deviceID, CL_DEVICE_HALF_FP_CONFIG,
+                                sizeof(value), &value, nullptr);
+        test_error(error, "Unable to get device CL_DEVICE_HALF_FP_CONFIG");
+        test_assert_error(value > 0,
+                          "CL_DEVICE_HALF_FP_CONFIG must return nonzero value");
+
+        test_assert_error((value & (CL_FP_ROUND_TO_NEAREST | CL_FP_INF_NAN))
+                              || (value & CL_FP_ROUND_TO_ZERO),
+                          "Reported half fp config doesn't meet minimum set");
+
+        error = clGetDeviceInfo(deviceID, CL_DEVICE_PREFERRED_VECTOR_WIDTH_HALF,
+                                sizeof(value), &value, nullptr);
+        test_error(
+            error,
+            "Unable to get device CL_DEVICE_PREFERRED_VECTOR_WIDTH_HALF");
+        test_assert_error(value > 0,
+                          "CL_DEVICE_PREFERRED_VECTOR_WIDTH_HALF must return "
+                          "nonzero value");
+
+        error = clGetDeviceInfo(deviceID, CL_DEVICE_NATIVE_VECTOR_WIDTH_HALF,
+                                sizeof(value), &value, nullptr);
+        test_error(error,
+                   "Unable to get device CL_DEVICE_NATIVE_VECTOR_WIDTH_HALF");
+        test_assert_error(
+            value > 0,
+            "CL_DEVICE_NATIVE_VECTOR_WIDTH_HALF must return nonzero value");
+    }
+    else
+    {
+        error = clGetDeviceInfo(deviceID, CL_DEVICE_HALF_FP_CONFIG,
+                                sizeof(value), &value, nullptr);
+        test_error(error, "Unable to get device CL_DEVICE_HALF_FP_CONFIG");
```

Comment:
This requires implementations without the `cl_khr_fp16` extension to recognise `CL_DEVICE_HALF_FP_CONFIG`, but support for `CL_DEVICE_HALF_FP_CONFIG` is only required if that extension is available, is it not? https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_HALF_FP_CONFIG says "provided by the cl_khr_fp16 extension". Our implementation, when FP16 support is disabled, returns `CL_INVALID_VALUE` here rather than setting `value` to `0`, and I think that is correct, since the spec also says it returns:

> [CL_INVALID_VALUE](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_INVALID_VALUE) if param_name is not one of the supported values or if size in bytes specified by param_value_size is < size of return type as specified in the [Device Queries](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#device-queries-table) table and param_value is not a NULL value or if param_name is a value that is available as an extension and the corresponding extension is not supported by the device.

Reply:
@hvdijk I agree with your analysis.  `CL_DEVICE_HALF_FP_CONFIG` is provided by the `cl_khr_fp16` extension, so if that extension is not supported the query should return `CL_INVALID_VALUE` indeed.

(the half query seems different from `CL_DEVICE_DOUBLE_FP_CONFIG`, which does not have a "provided by" clause)

Reply:
Thanks, created https://github.com/KhronosGroup/OpenCL-CTS/pull/2108 to change that.